Sources/StreamChatUI/Utils/AudioPlayer/AssetPropertyLoading.swift

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import AVFoundation
+import Foundation
+
+/// Defines a type that represents the properties of an asset that can be loaded
+public enum AssetProperty: String {
+    case duration
+}
+
+/// A simple type that describes the domain specific errors an instance of AssetPropertyLoading can return
+public enum AssetPropertyLoadingError: LocalizedError, Equatable {
+    case unknown
+    case cancelled
+
+    public var errorDescription: String? {
+        switch self {
+        case .unknown:
+            return "[\(type(of: self))] Asset's property loading failed with due to an unknown error"
+        case .cancelled:
+            return "[\(type(of: self))] Asset's property loading was cancelled"
+        }
+    }
+}
+
+/// A protocol that describes an object that can be used to load properties from an AVURLAsset
+public protocol AssetPropertyLoading {
+    /// A method that loads the property of an AVURLAsset asynchronously and
+    /// returns a result through a completion handler
+    ///
+    /// - Parameters:
+    /// - property: The property to load
+    /// - asset: The asset on which we will try to load the provided property
+    /// - onSuccessTransformer: If the load succeeds then the onSuccessTransformer
+    /// will be called, to transform the asset to result value we expect
+    /// - completion: The closure to call when we have a final result to
+    /// report (success or failure)
+    func loadProperty<Value>(
+        _ property: AssetProperty,
+        of asset: AVURLAsset,
+        onSuccessTransformer: @escaping (AVURLAsset) -> Value,
+        completion: @escaping (Result<Value, Error>) -> Void
+    )
+}
+
+public struct StreamAssetPropertyLoader: AssetPropertyLoading {
+    public func loadProperty<Value>(
+        _ property: AssetProperty,
+        of asset: AVURLAsset,
+        onSuccessTransformer: @escaping (AVURLAsset) -> Value,
+        completion: @escaping (Result<Value, Error>) -> Void
+    ) {
+        asset.loadValuesAsynchronously(forKeys: [property.rawValue]) {
+            handlePropertyLoadingResult(
+                property,
+                of: asset,
+                onSuccessTransformer: onSuccessTransformer,
+                completion: completion
+            )
+        }
+    }
```

Comment:
That was a tricky one. So for audio the duration is basically what we need. My idea was to implement just the bare minimum but after your feedback i tried to implement it and i realised that this will actually require work and potentially a change on the API. So, even if now we don't need it (and we may never really use it) i updated the code to load an array of properties.